Redesign classes to allow split deployments

ekohl opened the pull request:
Every class that provides an aspect of the deployment is now standalone. This means:

* candlepin
* pulp
* qpid
* application

They can just be included and work by themselves. All class parameters are tunables that users will commonly modify.

There's also shared parameters which are common among the standalone classes. These are meant to be changed via katello::params. Either via Hiera or by including them in a profile class before the instance classes themselves.

The katello::globals class has globals that a user via the installer might want to change. The katello::params class is not intended to be exposed via the installer.

Currently a draft because I want to do testing with rewriting the changes in the installer. Having this in a PR makes that easier.

Includes:
* https://github.com/theforeman/puppet-katello/pull/306
* https://github.com/theforeman/puppet-katello/pull/307 (only the drop unused variable)

ehelms commented:
I'm trying to follow the over arching design goals of this change and I am having trouble. I think what I need to understand more is the larger goal that we're driving towards. I have for a while thought the end goals would be something akin to:

 1) Make Katello like any other plugin in puppet-foreman
 2) Center backend configuration into either puppet-foreman_proxy_content or the installer
 3) Ensure we don't break user workflows or over complicate orchestrated configuration

ekohl commented:
Rebased so it's easier to see. Also submitted https://github.com/theforeman/puppet-katello/pull/309 as something that can be merged beforehand.

This is not final yet because I think I can rewrite init.pp to keep backwards compatible while already delivering the new split classes. That should allow experimentation with split deployments while still doing a monolithic deployment in the installer.

ekohl commented:
> This is not final yet because I think I can rewrite init.pp to keep backwards compatible while already delivering the new split classes. That should allow experimentation with split deployments while still doing a monolithic deployment in the installer.

This is done now. I've still added the monolithic change on top of it, but viewing the change should be easy now. It does drop some parameters, but I'd argue they didn't make sense for most users anyway.

ekohl commented:
I think this is now ready for review. It is backwards incompatible because it drops 6 parameters from `init.pp`:
* `$user`
* `$group`
* `$user_groups`
* `$qpid_interface`
* `$qpid_hostname`
* `$package_names`

I believe these shouldn't be configurable anyway so there's no loss of functionality here for most users.

@ehelms please have a look. Ideally we'd merge this soon so it has some time to stablize. It should also
@laugmanuel please also have a look. While it doesn't fully solve your use case yet, I'd like to get to a point where we provide suitable profiles (according to the [roles/profiles pattern](https://puppet.com/docs/pe/2019.2/the_roles_and_profiles_method.html)) that can be deployed standalone.

ekohl commented on manifests/params.pp:
```diff
@@ -1,86 +1,31 @@
+  Stdlib::Httpsurl $candlepin_url = "https://${facts['fqdn']}:8443/candlepin",
+  Stdlib::Httpsurl $pulp_url = "https://${facts['fqdn']}/pulp/api/v2/",
+  Stdlib::Httpsurl $crane_url = "https://${facts['fqdn']}:5000",
+  String $qpid_url = "amqp:ssl:${katello::globals::qpid_hostname}:5671",
```
I was also considering this. Given it was a parameter in `init.pp` before, it makes sense to keep it. Will have a look.

ekohl replied:
I've pushed a separate commit to make review easier but should be squashed on merge. Please have a look.

laugmanuel replied:
Change looks good to me.

ekohl commented on manifests/pulp.pp:
```diff
@@ -1,34 +1,91 @@
+#
+# @param repo_export_dir
+#   Create a repository export directory for Katello to use
 class katello::pulp (
```
My goal would for this PR is to just refactor without adding new params. Hopefully merge it quickly and then iterate on this in a separate PR.

laugmanuel replied:
Sounds good to me. If I can support you on getting the PR ready for merge just let me know.

laugmanuel replied:
Change looks good to me.

laugmanuel commented on manifests/init.pp:
```diff
@@ -40,6 +40,10 @@
 #
 # $qpid_wcache_page_size:: The size (in KB) of the pages in the write page cache
 #
+# $qpid_interface::     The interface qpidd listens to.
```
nitpick: `listens on`

laugmanuel commented:
@ekohl now we have some good news for you!

We have successfully upgraded our test environment as far as we can get it using the code in this PR.
What we cannot test at the moment is the upgrade of Pulp and Candlepin because we need the followup PR regarding Certs and `manage_httpd` (Candlepin is also affected by a different Cert locations / setup).

Next week I will configure a client to talk to the upgraded environment and test some things.

In my perspective this PR is ready.

ehelms commented on manifests/init.pp:
```diff
@@ -110,64 +102,117 @@
+    enable_deb    => $enable_deb,
+  }
 
+  class { 'katello::params':
```
Never seen calling params and configuring it as a design convention.

ekohl replied:
It's unconventional and I might clean this up a bit more later as I gain more experience with splitting up into cleaner roles and profiles.

ehelms commented:
What if any effect will this have on the installer currently? Just dropping a few parameters?

ekohl commented:
> What if any effect will this have on the installer currently? Just dropping a few parameters?

It should only drop a few parameters. Any other change would be considered a regression.

ehelms commented:
Will need a rebase, after that I think I am good with this change.

laugmanuel commented:
@ekohl I have one more thing regarding the Pulp configuration:
We do include `::foreman` here: https://github.com/theforeman/puppet-katello/pull/308/files#diff-f2ff10aff4c0e16a47fc381ed1a236f7R94
This basically results in the installation/configuration of a complete stack which does not make any sense if the components are split to different servers.

The only reason for including this class is to be able to access `$foreman::servername`. This variable is mapped to `$::fqdn` anyway, which is basicly the same as the default for `$pulp::server_name`.

What are your thoughts?

ekohl commented:
That's the hard part about our vhost handling. I'd say we need some handling that's roughly:
```puppet
if $manage_vhost {
  include foreman
  $servername = $foreman::servername
  foreman::config::apache::fragment { 'pulp':
    # ...
  }
} else {
  $servername = undef
}

class { 'pulp':
  servername   => $servername,
  manage_httpd => manage_vhost,
  # ...
}
```

laugmanuel commented:
> The only reason for including this class is to be able to access `$foreman::servername`. This variable is mapped to `$::fqdn` anyway, which is basicly the same as the default for `$pulp::server_name`.
And to be able to use `foreman::config::passenger::fragment` of course.

ekohl commented:
Rebased to resolve the conflict. I've also moved `$katello::params::qpid_interface` to `$katello::qpid::interface` since it's only used in that class.

ekohl commented:
Thanks for the reviews. Merged now and I'm looking forward to iterating on this.

laugmanuel commented:
Thanks @ekohl !
I did already prepare the follow up and will open a PR as soon as I've tested everything.
